src/cloud-api-adaptor/pkg/util/netops/netops.go

```diff
@@ -776,6 +780,139 @@ func (ns *namespace) RuleList(rule *Rule) ([]*Rule, error) {
 	return rules, nil
 }
 
+type Neighbor struct {
+	IP           netip.Addr
+	HardwareAddr string
+	Dev          string
+	State        NeighborState
+}
+
+type NeighborState uint16
+
+const NEIGHBOR_STATE_PERMANENT = netlink.NUD_PERMANENT
+
+var neighorStateMap = map[NeighborState]string{
+	netlink.NUD_NONE:       "none",
+	netlink.NUD_INCOMPLETE: "incomplete",
+	netlink.NUD_REACHABLE:  "reachable",
+	netlink.NUD_STALE:      "stale",
+	netlink.NUD_DELAY:      "delay",
+	netlink.NUD_PROBE:      "probe",
+	netlink.NUD_FAILED:     "failed",
+	netlink.NUD_NOARP:      "noarp",
+	netlink.NUD_PERMANENT:  "permanent",
+}
+
+func (s NeighborState) String() string {
+
+	str, ok := neighorStateMap[s]
+	if !ok {
+		return "unknown"
+	}
+	return str
+}
+
+var neighborStates = initLookupTable(maps.Keys(neighorStateMap))
+
+func (s NeighborState) MarshalJSON() ([]byte, error) {
+	return json.Marshal(s.String())
+}
+
+func (s *NeighborState) UnmarshalJSON(data []byte) error {
+
+	v, err := unmarshalJSON("neighbor state", data, neighborStates)
+	if err != nil {
+		return err
+	}
+	*s = v
+	return nil
+}
+
+// NeighborAdd adds a new neighbor entry
+
+func (ns *namespace) NeighborAdd(neighbor *Neighbor) error {
+
+	if !neighbor.IP.IsValid() {
+		return fmt.Errorf("failed to add a new neighbor entry: invalid IP address %q", neighbor.IP)
+	}
+
+	if neighbor.Dev == "" {
+		return fmt.Errorf("failed to add a new neighbor entry: no device name")
+	}
+
+	link, err := ns.handle.LinkByName(neighbor.Dev)
+	if err != nil {
+		return fmt.Errorf("failed to add a new neighbor entry: device %s not found: %w", neighbor.Dev, err)
+	}
+
+	hwAddr, err := net.ParseMAC(neighbor.HardwareAddr)
+	if err != nil {
+		return fmt.Errorf("failed to add a new neighbor entry: invalid hardware address %q", neighbor.HardwareAddr)
+	}
+
+	state := neighbor.State
+	if state.String() == "none" || state.String() == "unknown" {
```

Comment:
I should have said early: I'm not a networking expert at all. I was looking at the manual https://man7.org/linux/man-pages/man8/ip-neighbour.8.html and was wondering if we should attempt to add neighbors which state seems to indicate a transition like `incomplete` or `failed` (or maybe add the neighbor but let `state` unset).

Reply:
For our use case, I think we only need to identify and set permanent neighbor entries that are set statically. The other states are dynamically set when ARP packets are received. ARP packets are delivered to the podns network namespace in a peer pod VM via an VXLAN tunnel, so agent-protocol-forwarder does not need to manipulate those neighbor entries explicitly.